File: code/py/src/modular_api/core/error_response_middleware.py
from __future__ import annotations

from starlette.requests import Request
from starlette.responses import JSONResponse
from starlette.types import ASGIApp, Receive, Scope, Send

from modular_api.core.logger.logging_middleware import LOGGER_STATE_KEY


def error_response_middleware() -> type:
    class _ErrorResponseMiddleware:
        def __init__(self, app: ASGIApp) -> None:
            self.app = app

        async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
            if scope["type"] != "http":
                await self.app(scope, receive, send)
                return

            try:
                await self.app(scope, receive, send)
            except Exception as exc:
                request = Request(scope, receive, send)
                logger = getattr(request.state, LOGGER_STATE_KEY, None)
                if logger is not None:
                    logger.error(
                        "Unhandled error in request pipeline",
                        fields={"error": str(exc), "status": 500},
                    )
                response = JSONResponse({"error": "Internal server error"}, status_code=500)
                await response(scope, receive, send)

    return _ErrorResponseMiddleware

File: code/dart/modular_api/lib/src/core/error_response_middleware.dart
import 'dart:convert';

import 'package:shelf/shelf.dart';

import 'logger/logger.dart';
import 'logger/logging_middleware.dart';

Middleware errorResponseMiddleware() {
  const jsonHeaders = {'content-type': 'application/json; charset=utf-8'};

  return (Handler innerHandler) {
    return (Request request) async {
      try {
        return await innerHandler(request);
      } catch (error) {
        final logger = request.context[loggerContextKey] as ModularLogger?;
        logger?.error(
          'Unhandled error in request pipeline',
          fields: {'error': error.toString(), 'status': 500},
        );
        return Response(
          500,
          headers: jsonHeaders,
          body: jsonEncode({'error': 'Internal server error'}),
        );
      }
    };
  };
}